data_transfer/devices/dreem.py

```diff
@@ -8,7 +11,20 @@
 from data_transfer.db import all_filenames, create_record, read_record, update_record
 from data_transfer.lib import dreem as dreem_api
 from data_transfer.schemas.record import Record
-from data_transfer.services import inventory
+from data_transfer.services import inventory, ucam
+
+
+@dataclass
+class DreemRecording:
+    """
+    Stores most relevant metadata for readable lookup.
+    """
+
+    id: str
+    device_id: str
+    user_id: str
+    start: datetime
+    end: datetime
```

Comment:
In short, I have used the dataclass primarily for creating nice output to the logs otherwise the code becomes a bit of a mess. For reference, this is because dataclasses implement `[__repr__](https://docs.python.org/3/reference/datamodel.html#object.__repr__)` quite well. This outputs:

```json
Metadata cannot be determined for:
    DreemRecording(id='HASH_ID', device_id='HASH_ID', user_id='HASH_ID', start=datetime.datetime(2020, 12, 9, 21, 35), end=datetime.datetime(2020, 12, 10, 5, 38, 23)
```

> I understand the use of a data class here, and it helps in making its usages more strict whilst coding. Personally, I am not sure if I would use it in this case - as it is only used in the for item in unknown_records loop. As the dict is constructed via a 'private' method - it might actually make the translation between the Dreem API payload and our use a bit less transparent.

Note that `__recording_metadata` is essentially unpacking specific dictionary values from the complete metadata file, i.e., only the ones of interest. For example, previously the following was in the code:

```python
# Instead of using a dataclass:
recording = self.__recording_metadata(item)

# I had written:
record_id = recording["id"]
record_device_id = recording["device"]
record_user_id = recording["user"]
record_start = datetime.fromtimestamp(recording["report"]["start_time"])
record_end = datetime.fromtimestamp(recording["report"]["stop_time"])
```

Note that _logging_ takes place at multiple locations throughout this class, and critical to this is outputting all the above information. As sucess, previously we may have written:

```python
if not device_serial:
    # this would be duplicated throughout the code
    print(f"Unknown Device:\n   RecordingID: {record_id}, DeviceID: {record_device_id}")
    # Move onto next record: skips logic below to simplify error handling
    continue
```

With dataclasses, this becomes:

```python
if not device_serial:
    print(f"Unknown Device:\n   {recording}")
    # Move onto next record: skips logic below to simplify error handling
    continue
```

I agree that the intended use of this dataclass could be clearer. Shall I update a comment in the docstring and use of this? Also, do you agree/like the idea of using dataclasses in this way? This could be something that we do across all devices if desired.

Reply:
Thanks for the elaborate response Jay. Yes, it does make (more) sense now, especially with the simplicity for logging. 

How would you then (for example) approach the VTT dataclass? The API query only returns a limited set of attributes for the data class (compared to the Dreem, above). Basically only the `record_id` (which is the `vtt_hash`). Would you then update dataclass attributes as we uncover them, or only use the dataclass object for information directly from the manufacturer's API?

Reply:
> How would you then (for example) approach the VTT dataclass? The API query only returns a limited set of attributes for the data class (compared to the Dreem, above). 

In VTT's case we could have `id` and `date` (if we plan to use it)? In this case, a dataclass may feel inappropriate since it only has 1 or 2 values, but for consistency with other devices and logging it may be useful. 

> Basically only the record_id (which is the vtt_hash). Would you then update dataclass attributes as we uncover them, or only use the dataclass object for information directly from the manufacturer's API?

*I use the dataclass as a way to serialise the manufacturer's API to have more meaningful naming to make the code cleaner to read.*

Take Byteflies for example, you would have multiple values: `startDate`, `duration`, `id`, `groupId`, etc. The purpose of the data class was to create more meaningful names for these attributes. In the dreem implementation, the manufactuer's response has _a lot_ of metadata. We want to store it all, so although I construct the `DreemRecording` the complete object is not stored elsewhere and instead `item` is [saved](https://github.com/ideafast/middleware-services/blob/ft_ucam_dreem/data_transfer/devices/dreem.py#L132).

> If we do update the dataclass attributes as we uncover them, it is basically replacing the Record object. So if that is o.k. - we could also have the __recording_metadata() translate the Dreem payload straight into a Record object. thoughts?

This is a good point, however, the use of the dataclass is not _replacing_ the `Record` object but instead serialises the response from the manufacturer with only the relevant data needed to construct a Record. This makes (imho) reading the code easier, which is the key use of the dataclass here. Also note that the `item` metadata (manufactuer's response) is stored in `-meta.json` and not the dataclass.

Reply:
Got it, that makes senses. I'd say when we implement these across devices they'll become more obvious. Perhaps a few naming improvements would distinguish/clarify them a bit more - but overall I agree with your reasoning and the use of the different classes. Thanks for the elaborate explanation!